test/index.js

```diff
@@ -105,10 +108,48 @@ describe('KeyringController', function () {
       await keyringController.persistAllKeyrings();
 
       const { vault } = keyringController.store.getState();
-      const keyrings = await mockEncryptor.decrypt(password, vault);
+      const keyrings = await mockEncryptor.decrypt(PASSWORD, vault);
       expect(keyrings.indexOf(unsupportedKeyring) > -1).toBe(true);
       expect(keyrings).toHaveLength(2);
     });
+
+    describe('on a manifest v3 build', function () {
+      it('should save up to date encryption salt to the memStore', async function () {
+        keyringController.cacheEncryptionKey = true;
+        const vaultEncryptionKey = '🔑';
+        const vaultEncryptionSalt = '🧂';
+        const vault = JSON.stringify({ salt: vaultEncryptionSalt });
+        keyringController.store.updateState({ vault });
+
+        expect(keyringController.memStore.getState().encryptionKey).toBeNull();
+        expect(
+          keyringController.memStore.getState().encryptionSalt,
+        ).toBeUndefined();
+
+        await keyringController.unlockKeyrings(
+          undefined,
+          vaultEncryptionKey,
+          vaultEncryptionSalt,
+        );
+
+        expect(keyringController.memStore.getState().encryptionKey).toBe(
+          vaultEncryptionKey,
+        );
+        expect(keyringController.memStore.getState().encryptionSalt).toBe(
+          vaultEncryptionSalt,
+        );
+
+        const response = await keyringController.persistAllKeyrings();
```

Comment:
It looks like this is testing the case in `persistAllKeyrings` where `password` is unset but `encryptionKey` is set. While it's great to have coverage of this scenario, the one that was causing the extension bug is the case where `password` **is** set.

We could get to that state in one of two ways: either call `createNewVaultAndKeychain` or `submitPassword` during the "prepare" step of the test. After either of those calls, `persistAllKeyrings` should go down the path we've changed in this PR.

Reply:
It'd be nice to add a test case under `createNewVaultAndKeychain` as well, to confirm that the encryption key and salt are both saved. Likewise for `createNewVaultAndRestore`